pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -0,0 +1,311 @@
+func (m *mergeQuerier) LabelNames() ([]string, storage.Warnings, error) {
+	labelNames, warnings, err := m.mergeDistinctStringSlice(func(q storage.Querier) ([]string, storage.Warnings, error) {
+		return q.LabelNames()
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// check if the tenant label exists in the original result
+	var tenantLabelExists bool
+	labelPos := sort.SearchStrings(labelNames, defaultTenantLabel)
+	if labelPos < len(labelNames) && labelNames[labelPos] == defaultTenantLabel {
+		tenantLabelExists = true
+	}
+
+	labelToAdd := defaultTenantLabel
+
+	// if defaultTenantLabel already exists, we need to add the
+	// originalDefaultTenantLabel
+	if tenantLabelExists {
+		labelToAdd = originalDefaultTenantLabel
+		labelPos = sort.SearchStrings(labelNames, labelToAdd)
+	}
+
+	// insert label at the correct position
+	labelNames = append(labelNames, "")
+	copy(labelNames[labelPos+1:], labelNames[labelPos:])
+	labelNames[labelPos] = labelToAdd
+
+	return labelNames, warnings, nil
+}
+
+type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
+
+// mergeDistinctStringSlice is aggregating results from stringSliceFunc calls
+// on a querier. It removes duplicates and sorts the result. It doesn't require
+// the output of the stringSliceFunc to be sorted, as results of LabelValues
+// are not sorted.
+//
+// TODO: Consider running stringSliceFunc calls concurrently
+func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+	var warnings storage.Warnings
+	resultMap := make(map[string]struct{})
+	for pos, tenantID := range m.tenantIDs {
+		result, resultWarnings, err := f(m.queriers[pos])
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, e := range result {
+			resultMap[e] = struct{}{}
+		}
+		for _, w := range resultWarnings {
+			warnings = append(warnings, fmt.Errorf("error querying tenant id %s: %w", tenantID, w))
+		}
+	}
+
+	var result = make([]string, 0, len(resultMap))
+	for e := range resultMap {
+		result = append(result, e)
+	}
+	sort.Strings(result)
+	return result, warnings, nil
+}
+
+// Close releases the resources of the Querier.
+func (m *mergeQuerier) Close() error {
+	errs := tsdb_errors.NewMulti()
+	for pos, tenantID := range m.tenantIDs {
+		if err := m.queriers[pos].Close(); err != nil {
+			errs.Add(errors.Wrapf(m.queriers[pos].Close(), "failed to close querier for tenant id %s", tenantID))
+		}
+	}
+	return errs.Err()
+}
+
+// Select returns a set of series that matches the given label matchers. If the
+// tenantLabelName is matched on it only considers those queriers matching. The
+// forwarded labelSelector is not containing those that operate on
+// tenantLabelName.
+func (m *mergeQuerier) Select(sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
+	matchedTenants, filteredMatchers := filterValuesByMatchers(string(defaultTenantLabel), m.tenantIDs, matchers...)
```

Comment:
Nit: no need to typecast `string(defaultTenantLabel)`

Reply:
Good point: e51291b2678580be88bfd2a24554bfcfd4ac13c2